### ISSUE TITLE
HttpUrlConnectionExecutor user-agent

### ISSUE THREAD

**Issue by dmfs** [dmfs/http-client-essentials-suite#1]
`HttpUrlConnectionExecutor` should add the system default user-agent.
If the header is already present is should append its own user-agent to the present value.
It should append what `System.getProperty( "http.agent" )` returns.

**Comment by lemonboston** [comment]
Just copying here the relevant parts from the discussion from the PR:

> Sorry, I missed out that I was referring to decorating the HttpRequestExecutor, not HttpUrlConnection.
> So just as with Branded, only with appending at the end.
> Current (on master) HttpUrlConnectionExecutor would be responsible purely for adapting dmfs Http API to HttpUrlConnection, without modifying any value.
> And user-agent header appending would be done with decoration. The same approach as you envisioned with all these decorators, that separate aspects can be applied separately, no need to put everything into one class.
> Benefit would be that header append would be done with our API, so it can be re-used if other http client lib adapter executors are created later.
> 
> Client would not see this, decoration could be done in constructor, something like this:
> 
>     public HttpUrlConnectionExecutor(final HttpUrlConnectionFactory connectionFactory)
>     {
>         Product httpClientProduct = new VersionedProduct(new SafeToken(BuildConfig.NAME), new SafeToken(BuildConfig.VERSION));
>         Product systemHttpAgentProduct = new SafeProduct(System.getProperty("http.agent"));
>         mExecutor = new Branded(new Branded(new HttpUrlConnectionAdapterExecutor(connectionFactory), systemHttpAgentProduct, true), httpClientProduct, true);
>     }
> 
>     @Override
>     public <T> T execute(final URI uri, final HttpRequest<T> request)            
>     {
>         return mExecutor.execute(uri, request);        
>     }
> (HttpUrlConnectionAdapterExecutor is what HttpUrlConnectionExecutor is currently (on master))
> Or we could have a static factory method, like 
> static HttpRequestExecutor HttpUrlConnectionExecutorFactory.create().
> As I see it, in general, static methods should be avoided in favor of instance ones but with this sort of entry point to a library, they can actually help decoupling, because getting an instance won't be tied to a certain constructor of a certain subclass, future releases can do whatever they want under this factory method. This means we kind of draw a boundary with this for this module, this is the point where clients should get their executor instance that uses HttpUrlConnection. This can allow the flexibility to add anything later, if needed (like other decorators).
> (It's actually exactly what Factory pattern's intent is about as in here.)
> 
> For Branded, the extra boolean param to indicate appending in the end could be added, while keeping the other constructor that defaults to append in the beginning, and documenting it well. Or if we want to hide this usage from clients (although, I am not sure why is that needed, maybe they would need this to implement, modify their executor adapters of their own), then we could make a separate Branded with different name (there are just a few lines in it anyway).
> 
> We would need SafeProduct in this case for the system agent string, it would replace/escape characters as SafeComment does. Or RawStringProduct which would not do anything. (Currently we don't make this string safe either.)
> 
> Separating API adaption and agent-header append this way would make the testing easier as well.
> 
> What do you think?
> 
> (I think this would be a better way, but current version on this branch is working as well, so if you want to just make progress, we could merge this after review, and create separate enhancement issue. But I am happy to implement it next if you agree with the change.)
--lemonboston

--------

> Good one. You're completely right. Just a few thoughts:
> 
> We'll go with two executors:
> An unbranded one that doesn't add its own user agent (essentially the one we already have)
> A branded one that decorates the unbranded one and just delegates to that
> Both executors will be public, so users can choose (again, there might be a reason to have your client operating in "stealth mode", may it be for privacy reasons).
> Instead of using SafeProduct or something like RawStringProduct for the platform user agent, we would have a Platform which implements Product and takes no parameters (because it reads http.agent internally). So it would look like
> Product systemHttpAgentProduct = new Platform();
> I think we don't have to validate the platform user-agent. We trust that the platform returns a valid value.
> 
> We're not going to add that boolean parameter to Branded. The new decorator should be "internal" to HttpUrlConnectionExecutor. I want to make sure that users do not add their user-agent to the end. Only Executors should to that. Maybe we should create a module http-client-executor-utils with tools that only Executors need.
--dmfs

**Comment by lemonboston** [comment]
Thanks, I like this Platform class, but unfortunately, I missed to think through this system http agent handling fully with this approach when I suggested this. The problem is that Platform would not comply with the Product interface, because we wouldn't have name(), version(), comment(). This maybe tells us that it shouldn't be handled with our api after all, or that Product actually doesn't need these methods. They are not used anywhere currently, so we could actually just think of Product as a string, similarly to Token and Comment.

So the options I see for Platform:
1. Keep Product as it is, Platform throws UnsupportedOperationException from name(), version(), comment()
2. Keep Product as it is, Platform returns null from the methods (change name() contract, so it can be null)
3. Keep Product as it is, Platform implements a simple, fail-safe, lazy parsing of the http.agent in the class and return those values (or nulls)
4. Change Product, remove name(), version(), comment() as they are not needed

What do you think?
If you can foresee that we don't need to parse user-agent, and access the fields separately afterwards, or during request decoration (let's say subsequent decorator would do something based on product.name()), I would tend towards 4., because it could then also mean removing an unnecessary complexity from code.

----
About the two versions of HttpUrlConnectionExecutor: as I understand they would look something like this with the constructors:

```
class HttpUrlConnectionExecutor implements HttpRequestExecutor
{
    public HttpUrlConnectionExecutor()
    {
        this(new Finite(new DefaultHttpUrlConnectionFactory()));
    }

    public HttpUrlConnectionExecutor(final HttpUrlConnectionFactory connectionFactory)
    {
        mConnectionFactory = connectionFactory;
    }

class BrandedHttpUrlConnectionExecutor implements HttpRequestExecutor
    public BrandedHttpUrlConnectionExecutor()
    {
        .....
        mExecutor = new BottomBranded(new BottomBranded(new HttpUrlConnectionExecutor(), systemHttpAgentProduct), httpClientProduct);
    }

    public BrandedHttpUrlConnectionExecutor(HttpUrlConnectionFactory connectionFactory)
    {
        .....
        mExecutor = new BottomBranded(new BottomBranded(new HttpUrlConnectionExecutor(connectionFactory), systemHttpAgentProduct), httpClientProduct);
    }
```

I still think that a factory could be appropriate here. Because these constructors are basically configuring, setting up an instance of HttpRequestExecutor that uses HttpUrlConnection, and the more stuff we would add, the less straightforward would it be which constructor and what arguments to use for the client if he simply wants a 'default' instance.

I think the intent of factory when expressed as "provide a way for the client to create an object without exposing the instantiation logic and the actual subtype", applies here. So something like this could be more user-friendly:

```
public final class HttpUrlConnectionExecutorFactory
{

    /**
     * Create a HttpRequestExecutor that uses HttpUrlConnection as the underlying http library.
     * This default configuration uses timeouts ....  and adds the ... and system user-agent headers.
     * 
     * @return
     */
    public static HttpRequestExecutor createDefault()
    {
        HttpUrlConnectionFactory urlConnectionFactory = new Finite(new DefaultHttpUrlConnectionFactory());
        Product platform = new Platform();
        Product httpClientProduct = ...
        return new BottomBranded(new BottomBranded(new HttpUrlConnectionExecutor(urlConnectionFactory), platform), httpClientProduct);
    }
}
```
There could be another method for stealth mode if that was common usage, but if clients need customization they can still 'lego' their executor together anyway they want, the factory could also show a template for that.
This would allow to have only one class - HttpUrlConnectionExecutor - with one constructor instead of the 2 classes and 4 constructors above.
Basically this would separate configuration from the pure implementations, which is similar with the DI approach, but the motivation is a bit different here.

What do you say?

**Comment by dmfs** [comment]
Good questions. Initially I was leaning towards option 1. Even if we don't need these methods now, they exist for a good reason. If we decide to remove them we might have a breaking API change if we're going to re-add them later on (and we have a breaking API change now). Also, the `Platform` class is only for internal use, so it probably doesn't matter if these methods throw an exception. 
However, option 4 is probably fine as well. It depends on how likely it is that we or anyone else ever implement the `Product` interface outside of these libraries. If the existing classes cover all use cases and no one uses these methods we're fine. Since the current use cases probably don't require any custom implementations and since we don't parse anything, there is no point in calling these methods and we should be fine with removing them. Please go ahead and do so.

Regarding the classes. The branded class constructors should actually look like this:

```java
class BrandedHttpUrlConnectionExecutor implements HttpRequestExecutor
{
    public BrandedHttpUrlConnectionExecutor()
    {
       this(new HttpUrlConnectionExecutor());
    }

    public BrandedHttpUrlConnectionExecutor(HttpUrlConnectionFactory connectionFactory)
    {
      this(new HttpUrlConnectionExecutor(connectionFactory));
    }


    private BrandedHttpUrlConnectionExecutor(HttpRequestExecutor executor)
    {
        mExecutor = new BottomBranded(
                        new BottomBranded(executor, new Platform()),
                        new VersionedProduct(BuildConfig.NAME, BuildConfig.VERSION));
    }
```

And maybe we  swap the name and call the branded version just `HttpUrlConnectionExecutor` and the unbranded version `NakedHttpUrlConnectionExecutor` (or `PlainHttpUrlConnectionExecutor` in case anyone feels offended by the name).

I prefer this over the static factory method.

In another project I have a `DefaultExecutor` class that just delegates to another executor after applying several other default decorators. If you're happy with the defaults you just instantiate `new DefaultExecutor()` and you're good to go.

**Comment by lemonboston** [comment]
Thanks, implemented as suggested, so removed the methods from `Product `and used those constructors for the new executor.

PR is ready for review with these notes, questions:
- Platform now falls back to a "(no platform info)" comment in the probably rare case when `http.agent` property is not present. Is it okay like this?
- `BottomBranded `executor decorator only appends the user-agent headers if it exists already. Okay or should it always do it?
- New classes are in `executor_utils` sub-package now, should I go forward and create the mentioned http-client-executor-utils module?
If I should, is the following description okay?: "_Commonly used classes for HttpRequestExecutor implementations_". And the META-INF folder and its content - are they generated somehow, or okay to just copy them and edit the first 2 lines of the NOTICE?

**Comment by lemonboston** [comment]
I possibly messed the branching-merging-PRs up a bit, not sure what should the correct workflow of resolving problems with master...  let me know if you have a clear idea on this what is the best way.

@dmfs please review the PR #24 above, the only changes since your last review:
- revert `misc.xml` changes
- merging with master, resolving compilation issue with new Decorator name (`BottomBranded `line 54), and resolving IntelliJ build with module dependency updates in .iml files

**Comment by lemonboston** [comment]
By the way, this Travis integration is really great, that it can show that merged PR build would fail before actually merging.
